cpp/include/cudf/io/functions.hpp

```diff
@@ -238,6 +239,98 @@ table_with_metadata read_csv(
     read_csv_args const& args,
     rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+
+/**
+ * @brief Settings to use for `write_csv()`
+ */
+struct write_csv_args: private detail::csv::writer_options {
+  //inject private base accessors:
+  //
+  using detail::csv::writer_options::na_rep;
+  using detail::csv::writer_options::include_header;
+  using detail::csv::writer_options::rows_per_chunk;
+  using detail::csv::writer_options::line_terminator;
+  using detail::csv::writer_options::inter_column_delimiter;
+  using detail::csv::writer_options::true_value;
+  using detail::csv::writer_options::false_value;
+  
+  write_csv_args(sink_info const& snk,
+                 table_view const& table,
+                 std::string const& na,
+                 bool include_header,
+                 int rows_per_chunk,
+                 std::string line_term = std::string{"\n"},
+                 char delim = ',',
+                 std::string true_v = std::string{"true"},
+                 std::string false_v = std::string{"false"},
+                 table_metadata const* metadata = nullptr)
+    : writer_options(na,
+                     include_header,
+                     rows_per_chunk,
+                     line_term,
+                     delim,
+                     true_v,
+                     false_v),
+      sink_(snk),
+      table_(table),
+      metadata_(metadata)
+  {}
+
+  detail::csv::writer_options const& get_options(void) const
+  {
+    return *this;//sliced to base
+  }
+
+  sink_info const& sink(void) const
+  {
+    return sink_;
+  }
+
+  table_view const& table(void) const
+  {
+    return table_;
+  }
+
+  table_metadata const* metadata(void) const
+  {
+    return metadata_;
+  }
+
+private:
+  // Specify the sink to use for writer output:
+  //
+  sink_info const& sink_;
+
+  // Set of columns to output:
+  //
+  table_view const& table_;
+
+  // Optional associated metadata
+  //
+  table_metadata const* metadata_;
+
+};
+
+/**
+ * @brief Writes a set of columns to CSV format
+ *
+ * The following code snippet demonstrates how to write columns to a file:
+ * @code
+ *  #include <cudf.h>
+ *  ...
+ *  std::string filepath = "dataset.csv";
+ *  cudf::write_csv_args args{cudf::sink_info(filepath), table->view()};
+ *  ...
+ *  cudf::write_csv(args);
+ * @endcode
+ *
+ * @param args Settings for controlling writing behavior
+ * @param mr Optional resource to use for device memory allocation
+ */
+void write_csv(write_csv_args const& args, rmm::mr::device_memory_resource* mr =
```

Comment:
Since this function isn't returning any device memory, it doesn't need a resource.

Reply:
(1) all the other writer (ORC, PARQUET), and reader APIs pass a `device_memory_resource`; so, I'm just following a uniform API;
(2) I need the `mr` because I pass it to subsequent calls in the implementation of `write_csv()`. For example: `concatenate()` and various `from_<type>()` converter, which take the memory resource as parameter;

Reply:
It looks like none of the writers (including ORC and PARQUET) actually use the `mr` parameter (rightfully so, as Jake pointed out that the writers don't return any device memory, so they _shouldn't_ be using this parameter anyway). This parameter should be removed from `make_writer` below; I suppose this could be a follow-up PR instead of tacking it on to this one, since it affects all of the writers https://github.com/rapidsai/cudf/blob/b9e695325a1fae81bc5f17696fc987c4038079bd/cpp/src/io/functions.cpp#L47-L66

Reply:
Internal calls to APIs that allocate temporary device memory should be using the default memory resource rather than an `mr` parameter.

Reply:
> I suppose this could be a follow-up PR instead of tacking it on to this one, since it affects all of the writers

Yes, refactoring `cuio` writers is beyond the scope of this PR. I've been repeatedly told to follow the orc/parquet API, so I did. Whether those are subject to enhancement or not is probably something to follow-up on. @vuule , @OlivierNV any comments on that?

Reply:
The unnecessary unused _mr parameters to the writers came up before. I think it may have been  kept in the API in case the writers would ever need to return some permanent gpu result (one can imagine writing to a sink that returns a huge byte array, for example).
IMO, it probably doesn't hurt that it's in the api, whether we want this parameter in write_csv or not is ok with me either way.

Reply:
Okay, so I think perhaps it's time to make a decision: do we keep the `_mr` into the API, or not? Or, please specify the desired signature of the API.

Reply:
I vote for leaving it as is (into the API) for consistency.

Reply:
> I vote for leaving it as is

I agree. Going once, going twice...

Reply:
It's fine for this PR to leave the API for the reasons already discussed (refactoring the orc and parquet writers would be beyond the scope of this PR), but I still think we should refine the writer API in a future PR.

> in case the writers would ever need to return some permanent gpu result (one can imagine writing to a sink that returns a huge byte array, for example)

I think it would be more appropriate for the `data_sink` constructor to have the `mr` parameter in that case. The writer is supposed to be decoupled from the `data_sink`, so it shouldn't need to know anything about what the sink may or may not be doing with memory.

Reply:
> I think it would be more appropriate for the `data_sink` constructor to have the `mr` parameter in that case.

IIRC, in the function-style api, the writer creates the data sink, so the `mr` parameter would have to remain in the `write_csv` api, though it could be hypothetically removed from the writer() class if the hypothetical data sink creates the hypothetical device buffers (as opposed to the hypothetical writer creating device bufefrs for the sink to write into) :)